### PR TITLE
Make editor override service only operate on untyped editor inputs

### LINKED ISSUE #124351: Editor overrides should operate on untyped editors only
Related to https://github.com/microsoft/vscode/issues/106012, the editor override service ideally:
* works on exclusively on `IResourceEditorInput | IUntitledTextResourceEditorInput | IResourceDiffEditorInput`

The would allow us to prevent this `createEditorInput` from running before the overrides are figured out:

https://github.com/microsoft/vscode/blob/f19843b0a3e508d613f5b1b43d0de50cd83a3ac8/src/vs/workbench/services/editor/browser/editorService.ts#L602-L602

(also in other places). This helps prevent leaks in editor inputs.

### PR DESCRIPTION
**Pull request by lramos15**
This PR fixes #124351

This is the initial ground work for making the override service only care about untyped inputs.

A few concerned areas

- We need the active editor in a few places so I cast it to something with a viewtype to probe that property. We discussed not adding a viewType / editorID property to IEditorInput but you will see this design pattern a few times
- Particularly around notebooks, I'm not sure what we want to do about the DiffEditorInput name and description. You will see I've placed placeholder strings for now.

### CODE REVIEW

File: src/vs/workbench/services/editor/browser/editorService.ts [hunk 1]
```diff
@@ -1352,7 +1352,7 @@ export class DelegatingEditorService implements IEditorService {
 	findEditors(resource: IResourceEditorInputIdentifier, group: IEditorGroup | GroupIdentifier): IEditorInput | undefined;
 	findEditors(arg1: URI | IResourceEditorInputIdentifier, arg2?: IEditorGroup | GroupIdentifier): readonly IEditorIdentifier[] | readonly IEditorInput[] | IEditorInput | undefined { return this.editorService.findEditors(arg1, arg2); }
 
-	createEditorInput(input: IResourceEditorInputType): IEditorInput { return this.editorService.createEditorInput(input); }
+	createEditorInput(input: IResourceEditorInputType): EditorInput { return this.editorService.createEditorInput(input); }
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:1355]: Why is this required?

File: src/vs/workbench/services/editor/common/editorOverrideService.ts [hunk 2]
```diff
@@ -61,6 +60,8 @@ export interface IEditorType {
 	readonly providerDisplayName: string;
 }
 
+export type UntypedEditorInput = IResourceEditorInput | IUntitledTextResourceEditorInput | IResourceDiffEditorInput;
```
Comment (bpasero) [inline src/vs/workbench/services/editor/common/editorOverrideService.ts:63]: Should this be a type exported from where `IResourceEditorInput` is also exported?

File: src/vs/workbench/services/editor/common/editorService.ts [hunk 3]
```diff
@@ -224,7 +225,7 @@ export interface IEditorService {
 	/**
 	 * Converts a lightweight input to a workbench editor input.
 	 */
-	createEditorInput(input: IResourceEditorInputType): IEditorInput;
+	createEditorInput(input: IResourceEditorInputType): EditorInput;
```
Comment (bpasero) [inline src/vs/workbench/services/editor/common/editorService.ts:228]: Why is this required?

File: src/vs/workbench/services/editor/common/editorOverrideService.ts [hunk 4]
```diff
@@ -91,15 +92,15 @@ export type RegisteredEditorOptions = {
 
 export type ContributedEditorInfo = {
 	id: string;
-	describes: (currentEditor: IEditorInput) => boolean;
+	describes: (override: string) => boolean;
```
Comment (bpasero) [inline src/vs/workbench/services/editor/common/editorOverrideService.ts:95]: Instead of `override` should it be just something like "editorId"? The fact that this is an override is only evident as editor options.

File: src/vs/workbench/services/editor/browser/editorOverrideService.ts [hunk 5]
```diff
@@ -625,6 +629,15 @@ export class EditorOverrideService extends Disposable implements IEditorOverride
 		}
 		return false;
 	}
+
+	private isDiffEditorInput(editor: UntypedEditorInput): editor is IResourceDiffEditorInput {
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorOverrideService.ts:633]: Maybe this utility method could be exported from the place where `IResourceDiffEditorInput` is introduced, I think I have similar code in other places so there is chances of reuse.

File: src/vs/workbench/services/editor/browser/editorOverrideService.ts [hunk 6]
```diff
@@ -625,6 +629,15 @@ export class EditorOverrideService extends Disposable implements IEditorOverride
...
+		const testInput = editor as IResourceDiffEditorInput;
+		return testInput.originalInput !== undefined && testInput.modifiedInput !== undefined;
+	}
+
+	private getEditorResource(editor: UntypedEditorInput): URI | undefined {
```
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorOverrideService.ts:638]: Should this be added to `EditorResourceAccessor` as an overload?

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by bpasero** [comment]
> Particularly around notebooks, I'm not sure what we want to do about the DiffEditorInput name and description. You will see I've placed placeholder strings for now

Can you elaborate? Isn't the notebook component providing a factory to create diff editor inputs and so can they not provide name and description?

> We need the active editor in a few places so I cast it to something with a viewtype to probe that property. We discussed not adding a viewType / editorID property to IEditorInput but you will see this design pattern a few times

Can you list the places where we still have to do this or will be doing this going forward? Trying to understand the use case.

File: src/vs/workbench/contrib/customEditor/browser/customEditors.ts [hunk 7]
```diff
@@ -135,20 +136,23 @@ export class CustomEditorService extends Disposable implements ICustomEditorServ
...
+			}
 			// We check before calling this call back that both resources are defined
-			const input = CustomEditorInput.create(this.instantiationService, subInput.resource!, editorID, group.id, { customClasses });
+			const input = CustomEditorInput.create(this.instantiationService, subInput.resource, editorID, group.id, { customClasses });
 			return input instanceof EditorInput ? input : undefined;
 		};
 
-		const modifiedOverride = createEditorForSubInput(editor.modifiedInput, editorID, 'modified');
-		const originalOverride = createEditorForSubInput(editor.originalInput, editorID, 'original');
+		const modifiedOverride = createEditorForSubInput(editor.modifiedInput, editorID, 'modified') ?? this.editorService.createEditorInput(editor.modifiedInput);
```
Comment (lramos15) [inline src/vs/workbench/contrib/customEditor/browser/customEditors.ts:152]: @bpasero This is why I needed to change createEditorInput to return an `EditorInput`. Since every implementation was returning one anyways I didn't feel it was a huge change.

File: src/vs/workbench/services/editor/browser/editorOverrideService.ts [hunk 8]
```diff
@@ -452,7 +455,8 @@ export class EditorOverrideService extends Disposable implements IEditorOverride
 		}
 		// Map the editors to quickpick entries
 		registeredEditors.forEach(editor => {
-			const isActive = currentEditor ? editor.editorInfo.describes(currentEditor) : false;
+			const currentViewType = (currentEditor as IContributedEditorInput).viewType ?? DEFAULT_EDITOR_ASSOCIATION.id;
```
Comment (lramos15) [inline src/vs/workbench/services/editor/browser/editorOverrideService.ts:458]: @bpasero Here's an example of that common casting pattern where I have an `IEditorInput` but need the viewType from it (even if it's undefined). So I cast it to an editor input that has a viewtype.

File: src/vs/workbench/services/editor/common/editorOverrideService.ts (hunk 2 above)
Comment (lramos15) [reply src/vs/workbench/services/editor/common/editorOverrideService.ts]: There was an existing type which I switched to using. `IResourceEditorInputType`

File: src/vs/workbench/services/editor/browser/editorOverrideService.ts (hunk 6 above)
Comment (lramos15) [reply src/vs/workbench/services/editor/browser/editorOverrideService.ts]: I've enriched the `EditorResourceAccessor` to handle this case.

**Comment by lramos15** [comment]
> Can you elaborate? Isn't the notebook component providing a factory to create diff editor inputs and so can they not provide name and description?

They could, I think the primary reason for using the one from the DiffEditor is to stay consistent with other diff editors. The custom editor for example just uses the normal `DiffEditorInput` with custom editor subinputs. This allows me to pass undefined and have the same diff editor logic generate the name as if none was provided. I'm not sure if `NotebookDiffEditorInput` could extend `DiffEditorInput` instead of `EditorInput` to receive that same level of functionality.

// cc @jrieken and @rebornix for possible context on how to move forward with that

File: src/vs/workbench/services/editor/browser/editorOverrideService.ts (hunk 8 above)
Comment (bpasero) [reply src/vs/workbench/services/editor/browser/editorOverrideService.ts:459]: I will reply in my review comment on this.

File: src/vs/workbench/common/editor.ts [hunk 9]
```diff
@@ -835,24 +841,30 @@ class EditorResourceAccessorImpl {
 	 * form so that only one editor opens for same file URIs with different casing. As
 	 * such, the original URI and the canonical URI can be different.
 	 */
-	getOriginalUri(editor: IEditorInput | undefined | null): URI | undefined;
-	getOriginalUri(editor: IEditorInput | undefined | null, options: IEditorResourceAccessorOptions & { supportSideBySide?: SideBySideEditor.PRIMARY | SideBySideEditor.SECONDARY }): URI | undefined;
-	getOriginalUri(editor: IEditorInput | undefined | null, options: IEditorResourceAccessorOptions & { supportSideBySide: SideBySideEditor.BOTH }): URI | { primary?: URI, secondary?: URI } | undefined;
-	getOriginalUri(editor: IEditorInput | undefined | null, options?: IEditorResourceAccessorOptions): URI | { primary?: URI, secondary?: URI } | undefined {
+	getOriginalUri(editor: IEditorInput | IResourceEditorInputType | undefined | null): URI | undefined;
```
Comment (bpasero) [inline src/vs/workbench/common/editor.ts:844]: Please update the existing `EditorResourceAccessor` tests to account for testing untyped inputs (normal, diff, untitled).

File: src/vs/workbench/contrib/files/browser/editors/binaryFileEditor.ts [hunk 10]
```diff
@@ -49,15 +49,29 @@ export class BinaryFileEditor extends BaseBinaryResourceEditor {
...
-				replacement: overridenInput?.editor ?? input,
-				options: {
-					...overridenInput?.options ?? options,
-					override: EditorOverride.DISABLED
-				}
-			}], overridenInput?.group ?? this.group);
+			if (overridenInput === OverrideStatus.ABORT) {
+				return;
+			} else if (overridenInput === OverrideStatus.NONE) {
+				// Replace the overrriden input, with the text based input
```
Comment (bpasero) [inline src/vs/workbench/contrib/files/browser/editors/binaryFileEditor.ts:55]: There is some repetition between this block and the next one. Can we extract the reusable thing (e.g. the call to `replaceEditors`) and only massage the input as needed?

File: src/vs/workbench/contrib/notebook/browser/notebookServiceImpl.ts [hunk 11]
```diff
@@ -165,12 +163,12 @@ export class NotebookProviderInfoStore extends Disposable {
...
 				return { editor: NotebookEditorInput.create(this._instantiationService, notebookUri, notebookProviderInfo.id), options: notebookOptions };
 			};
-			const notebookEditorDiffFactory: DiffEditorInputFactoryFunction = (diffEditorInput: DiffEditorInput, options, group) => {
+			const notebookEditorDiffFactory: DiffEditorInputFactoryFunction = diffEditorInput => {
 				const modifiedInput = diffEditorInput.modifiedInput;
 				const originalInput = diffEditorInput.originalInput;
 				const notebookUri = modifiedInput.resource!;
 				const originalNotebookUri = originalInput.resource!;
-				return { editor: NotebookDiffEditorInput.create(this._instantiationService, notebookUri, modifiedInput.getName(), originalNotebookUri, originalInput.getName(), diffEditorInput.getName(), notebookProviderInfo.id) };
+				return { editor: NotebookDiffEditorInput.create(this._instantiationService, notebookUri, 'Modified Input Name', originalNotebookUri, 'Original Input Name', 'Diff Editor Input Name', notebookProviderInfo.id) };
```
Comment (bpasero) [inline src/vs/workbench/contrib/notebook/browser/notebookServiceImpl.ts:171]: I think this can be solved by `NotebookDiffEditorInput` extending `DiffEditorInput` so that the name is properly computed. This seems just like an oversight that it isn't the case already.

File: src/vs/workbench/contrib/customEditor/browser/customEditors.ts [hunk 12]
```diff
@@ -117,7 +118,7 @@ export class CustomEditorService extends Disposable implements ICustomEditorServ
 						id: contributedEditor.id,
 						label: contributedEditor.displayName,
 						detail: contributedEditor.providerDisplayName,
-						describes: (currentEditor) => currentEditor instanceof CustomEditorInput && currentEditor.viewType === contributedEditor.id,
+						describes: editorId => editorId === contributedEditor.id,
```
Comment (bpasero) [inline src/vs/workbench/contrib/customEditor/browser/customEditors.ts:121]: It now seems to me that every implementation of `describes` is comparing whether the passed in `editorId` matches the contributed editor ID. Isn't that something we could do from the outside because the editor is already contributed with an identifier? Why does the contributor here need to make this check?

**Review by bpasero — CHANGES_REQUESTED** [review]
Left some feedback inline and have some follow up questions here:

**Override Status**
Can you elaborate on the intent of this new override status? If we end up needing this, can we have test coverage for that too?

**Probing on `EditorInput.viewType`**
I found a couple of places now that do `editor as as IContributedEditorInput` to probe on the `viewType`. Could we avoid doing that by looping over `RegisteredEditor` and maybe ask a method such as `RegisteredEditor.editoInfo.describes` passing in the `IEditorInput` like we did before and then the contributor can do the casting?

File: src/vs/workbench/contrib/customEditor/browser/customEditors.ts (hunk 12 above)
Comment (lramos15) [reply src/vs/workbench/contrib/customEditor/browser/customEditors.ts:121]: Good catch. It most likely is, let me go ahead and internalize this check

File: src/vs/workbench/contrib/notebook/browser/notebookServiceImpl.ts (hunk 11 above)
Comment (lramos15) [reply src/vs/workbench/contrib/notebook/browser/notebookServiceImpl.ts:171]: Yeah that's what my above comment said. I would have to ask Joh and Peng why they went with EditorInput.

**Comment by lramos15** [comment]
I'll answer your two main questions here.

**Override Status**
The override status is just to allow a mapping to the old way to prevent the failing tests and ensure the new behavior matches the old. If we completely internalize typed editor inputs to just the editor service and only allow passing of untyped editor inputs they might go away but that's a larger work item than this current PR.  The two statuses are `OverrideStatus.ABORT` and `OverrideStatus.NONE` 

ABORT -> User has cancelled the picker or explicitly done something in which no editor could be opened
NONE -> We don't have enough info to resolve an editor input for you. It is up to the calling function to decide what to do.

Previously, we would return the resolved `IEditorInput` or `undefined`. If we didn't have enough info to resolve the editor input we would just return the passed one back and then `undefined` would be the abort case. The problem with that approach now is that we have untyped editor inputs and calling `createEditorInput` on them caused all the test to fail even if overrides were disabled. 

**Probing on EditorInput.viewType**
This sounds like we're just moving the problem around. It would still be doing a cast and `describes` now takes an `editorId` not an `IEditorInput`. `describes` is also going away as all of them have become the same.

**Comment by bpasero** [comment]
> Override Status

Ok, if its temporary 👍 . Saw a bit of duplicated code in editor service for handling with the override result, but fine to leave it then.

> Probing on EditorInput.viewType

Another idea: what if `EditorInput` has a method `isOfType` (maybe better `matchesType(editorId: string)`?) and the default implementation is doing `editorId === this.editorId` but subclasses can override this to probe on the `viewType` if they have this concept?

**Comment by bpasero** [comment]
> We could introduce that function but we would still need to introduce the concept of an `editorId` at that point.

I am not against `editorId`, we already have `typeId` for that purpose. What I don't find ideal is having another concept called `viewTypeId` that seems like a secondary editor identification that only notebooks and custom editors use because they are driven by extensions.

File: src/vs/workbench/common/editor.ts (hunk 9 above)
Comment (lramos15) [reply src/vs/workbench/common/editor.ts:844]: Added another set of tests that does the same thing as the old ones but with untyped inputs.

**Review by bpasero — APPROVED** [review]
